### PR TITLE
Adds in discord auto-roling

### PR DESCRIPTION
**Pull request by AffectedArc07**
## About The Pull Request
Oranges asked for this

When a user successfully links their BYOND accounts and discord accounts, they will now be automatically roled, based on config settings. 

NOTE: This does not update pre-linked accounts, but I can write an external bot to do that if needed

RUST-G has also been updated in this PR, to be inline with this PR over at RUST-G https://github.com/tgstation/rust-g/pull/23

## Why It's Good For The Game
People with linked accounts can now be given a specific role in the discord, which is good for verification.

## Changelog
:cl: AffectedArc07
add: Linking your discord and BYOND accounts will now give you a role in the discord
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/discord.dm [hunk 1]
```diff
@@ -38,12 +38,13 @@ SUBSYSTEM_DEF(discord)
 	var/enabled = 0 // Is TGS enabled (If not we wont fire because otherwise this is useless)
 
 /datum/controller/subsystem/discord/Initialize(start_timeofday)
+	grant_role("200631029675982858")
```
Comment (skull132) [inline code/controllers/subsystem/discord.dm]: HMMMMMMMMMMM.

File: code/controllers/subsystem/discord.dm [hunk 2]
```diff
@@ -113,3 +113,11 @@ SUBSYSTEM_DEF(discord)
...
 	var/regex/num_only = regex("\[^0-9\]", "g")
 	return num_only.Replace(input, "")
+
+/datum/controller/subsystem/discord/proc/grant_role(id)
+	// Ignore this shit if config isnt enabled for it
+	if(!CONFIG_GET(flag/enable_discord_autorole))
+		return
+
+	// Make the request
+	rustg_http_request_blocking(RUSTG_HTTP_METHOD_PUT, "https://discordapp.com/api/guilds/[CONFIG_GET(string/discord_guildid)]/members/[id]/roles/[CONFIG_GET(string/discord_roleid)]", "",  "{\"Authorization\":\"Bot [CONFIG_GET(string/discord_token)]\"}")
```
Comment (optimumtact) [inline code/controllers/subsystem/discord.dm]: maybe add the request and response log here and put it into a separate log?

Suspect we'll probably need that at least once while configuring this.

File: code/controllers/subsystem/discord.dm [hunk 3]
```diff
@@ -113,3 +113,12 @@ SUBSYSTEM_DEF(discord)
...
 	return num_only.Replace(input, "")
+
+/datum/controller/subsystem/discord/proc/grant_role(id)
+	// Ignore this shit if config isnt enabled for it
+	if(!CONFIG_GET(flag/enable_discord_autorole))
+		return
+
+	// Make the request
+	var/response = rustg_http_request_blocking(RUSTG_HTTP_METHOD_PUT, "https://discordapp.com/api/guilds/[CONFIG_GET(string/discord_guildid)]/members/[id]/roles/[CONFIG_GET(string/discord_roleid)]", "",  "{\"Authorization\":\"Bot [CONFIG_GET(string/discord_token)]\"}")
+	WRITE_LOG(GLOB.discord_api_log, "DISCORD API: [response]")
```
Comment (optimumtact) [inline code/controllers/subsystem/discord.dm]: is it a json string?

Comment (optimumtact) [reply code/controllers/subsystem/discord.dm]: also please log the url as well :D

Comment (AffectedArc07) [reply code/controllers/subsystem/discord.dm]: Yes. Do you want each item of it to be newlined?

Comment (optimumtact) [reply code/controllers/subsystem/discord.dm]: no, just slap it in

**Comment by AffectedArc07** [comment]
DNM while I add in the script to role existing users

**Comment by optimumtact** [comment]
Awaiting rust-g release

**Comment by AffectedArc07** [comment]
Dont know why the bot kept DNM on here but it can do

File: code/controllers/subsystem/discord.dm [hunk 4]
```diff
@@ -113,3 +113,13 @@ SUBSYSTEM_DEF(discord)
...
 	return num_only.Replace(input, "")
+
+/datum/controller/subsystem/discord/proc/grant_role(id)
+	// Ignore this shit if config isnt enabled for it
+	if(!CONFIG_GET(flag/enable_discord_autorole))
+		return
+
+	var/url = "https://discordapp.com/api/guilds/[CONFIG_GET(string/discord_guildid)]/members/[id]/roles/[CONFIG_GET(string/discord_roleid)]"
+	// Make the request
+	var/response = rustg_http_request_blocking(RUSTG_HTTP_METHOD_PUT, url, "",  "{\"Authorization\":\"Bot [CONFIG_GET(string/discord_token)]\"}")
```
Comment (SpaceManiac) [inline code/controllers/subsystem/discord.dm]: hi isn't the whole point to use the async version

**Comment by optimumtact** [comment]
make sure you add the latest release version of the rust_g dll plox

**Comment by AffectedArc07** [comment]
@optimumtact done

File: code/controllers/subsystem/discord.dm (hunk 4 above)
Comment (optimumtact) [reply code/controllers/subsystem/discord.dm]: @AffectedArc07 last thing here?

**Comment by AffectedArc07** [comment]
Async is in, and here is an example of the logs.

```hs
[2020-02-20 09:26:49.776] Starting up round ID .
 - -------------------------
[2020-02-20 09:27:42.634] PUT https://discordapp.com/api/guilds/454650535036059669/members/200631029675982858/roles/678175729565040698 returned 204 
```

**Pull request merged**